### PR TITLE
DBZ-1169 Outbox support

### PR DESCRIPTION
**Pull request by renatomefi**
## Context
[DBZ-1169](https://issues.jboss.org/browse/DBZ-1169)

## This PR
This is a Draft PR which will try to reflect upon the decisions in the design of the outbox support by Debezium

Let's leave this PR for technical reviews, about the behavior let's centralize in [Jira](https://issues.jboss.org/browse/DBZ-1169).

## Todo
- [x] Configuration definition
  - [x] Configuration for Event Timestamp
  - [x] Configuration for extra fields translated to headers or message body
    - My idea here is to create a mapping where you can set the field and its destination: `event_id:header,my_extra_field:value`
  - [x] Validate configuration
- [x] Deal with unexpected operations
  - [x] Is `IllegalStateException` fine? We cannot throw `ConnectException` within `apply`
- [x] Ignore deletions and tombstones
- [x] Create Event Record with its values
- [x] Deal with Schema and Schemaless?
- [x] Route by field support
  - [x] Set record `Key` based on EventKey or fallback to AggregateId
- [x] Support Event Timestamp or fallback to `ts_ms`
- [x] Extra fields support (As header, payload or ignore)
- [x] Support Schema version
- [x] Integration tests

### CODE REVIEW

File: debezium-core/src/main/java/io/debezium/transforms/outbox/EventRouterConfigDefinition.java [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+
+    private static final Field FIELD_PAYLOAD_TYPE = Field.create("table.field.payload.type")
+            .withDisplayName("Event Payload Type Field")
+            .withType(ConfigDef.Type.STRING)
+            .withWidth(ConfigDef.Width.MEDIUM)
+            .withImportance(ConfigDef.Importance.LOW)
+            .withDefault("aggregatetype")
+            .withDescription("The field which contains the Payload Type within the outbox table");
+
+    private static final Field ROUTE_BY_FIELD = Field.create("route.by.field")
```
Comment (renatomefi) [inline debezium-core/src/main/java/io/debezium/transforms/outbox/EventRouterConfigDefinition.java]: Not enum, it's a string with the column name, default: "aggregatetype"

File: debezium-core/src/main/java/io/debezium/transforms/outbox/EventRouter.java [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+    private final ExtractField<R> afterExtractor = new ExtractField.Value<>();
+    private final RegexRouter<R> regexRouter = new RegexRouter<>();
+    private EventRouterConfigDefinition.InvalidOperationBehavior invalidOperationBehavior;
+
+    private String fieldEventId;
+    private String fieldEventKey;
+    private String fieldEventType;
+    private String fieldPayload;
+    private String fieldPayloadId;
+    private String fieldPayloadType;
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/transforms/outbox/EventRouter.java]: What is meaning of "payloadType"?

File: debezium-core/src/main/java/io/debezium/transforms/outbox/EventRouter.java [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+
+        Long timestamp = debeziumEventValue.getInt64("ts_ms");
+
+        String eventId = eventStruct.getString(fieldEventId);
+        String eventType = eventStruct.getString(fieldEventType);
+        String payload = eventStruct.getString(fieldPayload);
+        String payloadId = eventStruct.getString(fieldPayloadId);
+        String payloadType = eventStruct.getString(fieldPayloadType);
+
+        Schema valueSchema = SchemaBuilder.struct()
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/transforms/outbox/EventRouter.java]: Can we cache the schema? I remember building it can be costly from earlier SMTs.

File: debezium-core/src/test/java/io/debezium/transforms/outbox/EventRouterTest.java [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+        before.put("aggregatetype", payloadType);
+        before.put("aggregateid", payloadId);
+        before.put("type", eventType);
+        before.put("payload", payload);
+
+        final Struct body = envelope.create(before, null, System.nanoTime());
+        return new SourceRecord(new HashMap<>(), new HashMap<>(), "db.outbox", envelope.schema(), body);
+    }
+
+}
```
Comment (gunnarmorling) [inline debezium-core/src/test/java/io/debezium/transforms/outbox/EventRouterTest.java]: Minor nit: newline should be added at EOF.

File: debezium-core/src/main/java/io/debezium/transforms/outbox/EventRouter.java [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+
+        R newRecord = r.newRecord(
+                eventStruct.getString(routeByField).toLowerCase(),
+                null,
+                Schema.STRING_SCHEMA,
+                defineRecordKey(eventStruct, payloadId),
+                valueSchema,
+                value,
+                r.timestamp(),
+                null
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/transforms/outbox/EventRouter.java]: The SMT from the blog post propagated the event id (_not_ aggregate id) as a message header property, allowing consumers to detect duplicates efficiently, without having to parse the payload. Do you plan to add this here, too?

File: debezium-core/src/main/java/io/debezium/transforms/outbox/EventRouterConfigDefinition.java [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+            .withDefault("type")
+            .withDescription("The field which contains the Event Type within the outbox table");
+
+    static final Field FIELD_PAYLOAD = Field.create("table.field.payload")
+            .withDisplayName("Event Payload Field")
+            .withType(ConfigDef.Type.STRING)
+            .withWidth(ConfigDef.Width.MEDIUM)
+            .withImportance(ConfigDef.Importance.LOW)
+            .withDefault("payload")
+            .withDescription("The field which contains the Event Type within the outbox table");
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/transforms/outbox/EventRouterConfigDefinition.java]:
```suggestion
            .withDescription("The column which contains the payload within the outbox table");
```

Comment (gunnarmorling) [reply debezium-core/src/main/java/io/debezium/transforms/outbox/EventRouterConfigDefinition.java]: (should be "column" instead of "field" in all the option descriptions)

File: debezium-core/src/main/java/io/debezium/transforms/outbox/EventRouterConfigDefinition.java [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+            .withImportance(ConfigDef.Importance.LOW)
+            .withDefault("aggregateid")
+            .withDescription("The field which contains the Payload ID within the outbox table");
+
+    static final Field FIELD_PAYLOAD_TYPE = Field.create("table.field.payload.type")
+            .withDisplayName("Event Payload Type Field")
+            .withType(ConfigDef.Type.STRING)
+            .withWidth(ConfigDef.Width.MEDIUM)
+            .withImportance(ConfigDef.Importance.LOW)
+            .withDefault("aggregatetype")
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/transforms/outbox/EventRouterConfigDefinition.java]: Is this the correct default?

File: debezium-core/src/main/java/io/debezium/transforms/outbox/EventRouterConfigDefinition.java [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+                    " the topic name");
+
+    static final Field ROUTE_TOPIC_REGEX = Field.create("route.topic.regex")
+            .withDisplayName("The name of the routed topic")
+            .withType(ConfigDef.Type.STRING)
+            .withValidation(Field::isRegex)
+            .withDefault("(.*)")
+            .withWidth(ConfigDef.Width.MEDIUM)
+            .withImportance(ConfigDef.Importance.LOW)
+            .withDescription("The default regex to use within te RegexRouter, the default capture will allow" +
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/transforms/outbox/EventRouterConfigDefinition.java]:
```suggestion
            .withDescription("The default regex to use within the RegexRouter, the default capture will allow" +
```

File: debezium-core/src/main/java/io/debezium/transforms/outbox/EventRouterConfigDefinition.java [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+            .withImportance(ConfigDef.Importance.HIGH)
+            .withDescription("The name of the topic in which the events will be routed, a replacement '$1' is" +
+                    " available which is the value of the field configured via 'route.by.field'");
+
+    static final Field OPERATION_INVALID_BEHAVIOR = Field.create("debezium.op.invalid.behavior")
+            .withDisplayName("Behavior when the route fails to apply")
+            .withEnum(InvalidOperationBehavior.class, InvalidOperationBehavior.SKIP_AND_WARN)
+            .withWidth(ConfigDef.Width.MEDIUM)
+            .withImportance(ConfigDef.Importance.MEDIUM)
+            .withDescription("While debezium is monitoring the table, it's expecting only to see 'create' row events," +
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/transforms/outbox/EventRouterConfigDefinition.java]:
```suggestion
            .withDescription("While Debezium is monitoring the table, it's expecting only to see 'CREATE' and 'DELETE' row events," +
```

File: debezium-core/src/main/java/io/debezium/transforms/outbox/EventRouter.java (hunk 2 above)
Comment (renatomefi) [reply debezium-core/src/main/java/io/debezium/transforms/outbox/EventRouter.java]: So this is the "generic" name for the aggregatetype, the default value of this field is `aggregatetype` as from the outbox blog post

File: debezium-core/src/main/java/io/debezium/transforms/outbox/EventRouter.java (hunk 5 above)
Comment (renatomefi) [reply debezium-core/src/main/java/io/debezium/transforms/outbox/EventRouter.java]: Yes, it's on my to do list, I thought about treating this one as "extra fields", default going to header but with the option to be part of the event envelope. Wdyt?

File: debezium-core/src/main/java/io/debezium/transforms/outbox/EventRouterConfigDefinition.java (hunk 7 above)
Comment (renatomefi) [reply debezium-core/src/main/java/io/debezium/transforms/outbox/EventRouterConfigDefinition.java]: That's related to your previous question, beign the generic name for the aggregate type. Do you have another suggestion?
the name I took from your blogpost! hehe

File: debezium-core/src/main/java/io/debezium/transforms/outbox/EventRouterConfigDefinition.java (hunk 6 above)
Comment (renatomefi) [reply debezium-core/src/main/java/io/debezium/transforms/outbox/EventRouterConfigDefinition.java]: Also the static properties?

File: debezium-core/src/main/java/io/debezium/transforms/outbox/EventRouter.java (hunk 2 above)
Comment (gunnarmorling) [reply debezium-core/src/main/java/io/debezium/transforms/outbox/EventRouter.java]: Gotcha. Thanks for clarifying.

File: debezium-core/src/main/java/io/debezium/transforms/outbox/EventRouter.java (hunk 5 above)
Comment (gunnarmorling) [reply debezium-core/src/main/java/io/debezium/transforms/outbox/EventRouter.java]: I think just header is fine. We can always evolve it later on.

Comment (renatomefi) [reply debezium-core/src/main/java/io/debezium/transforms/outbox/EventRouter.java]: It's fixed, but I really need your validation on the approach of "extra fields"

File: debezium-core/src/main/java/io/debezium/transforms/outbox/EventRouterConfigDefinition.java (hunk 6 above)
Comment (gunnarmorling) [reply debezium-core/src/main/java/io/debezium/transforms/outbox/EventRouterConfigDefinition.java]: Depends on whether they refer to a column of a table or the field of a Kafka message (I'm not sure).

File: debezium-core/src/main/java/io/debezium/transforms/outbox/EventRouterConfigDefinition.java (hunk 7 above)
Comment (gunnarmorling) [reply debezium-core/src/main/java/io/debezium/transforms/outbox/EventRouterConfigDefinition.java]: No, it make sense, I just was missing the semantics at first. Thanks!

**Comment by gunnarmorling** [comment]
@renatomefi, just was reminded of the discussion with @jgao54 regarding a field for the schema version. It seems we wouldn't have to "bake this in", as it can be added by the user via that notion of "extra fields". WDYT?

**Comment by renatomefi** [comment]
> @renatomefi, just was reminded of the discussion with @jgao54 regarding a field for the schema version. It seems we wouldn't have to "bake this in", as it can be added by the user via that notion of "extra fields". WDYT?

It's a bit more complicated, I didn't forget about it but I was leaving it to the end because the version is part of the metadata and has to be set in the `SchemaBuilder` rather than a normal value, [check this](https://github.com/apache/kafka/blob/37efc79eb82e0f290d0d2a17f24001f2cbc922cc/connect/api/src/main/java/org/apache/kafka/connect/data/SchemaBuilder.java#L164).
So what I want to do is to create a Map of version and schema, since while capturing you might have multiple versions of the schema among your records, specially if your KC service is running for a while!
Is that approach valid?

Also, if you check the latest commit `DBZ-1169 Schema should inherit from original type`, what do you think about the approach to reuse the original schema? There's one caveat which is the field used for routing, this one should always be string since it becomes a topic, to ensure this behavior I'm using `getString` and the user will get an exception otherwise

**Comment by gunnarmorling** [comment]
Hey @renatomefi, I saw you pushed some more things to this PR. Do you have more planned or do you consider it done from your POV? Any specific things you feel we should discuss apart from a general review (planned for tomorrow, so barring any severe findings it could go into the next release)?

**Comment by renatomefi** [comment]
Hey @gunnarmorling, I think I'm done, I still want to increase tests for edge cases and misconfiguration but I can do that later, also I'll work on the docs during the week, so I'd say it's ready to go if you approve it =)

File: debezium-core/src/main/java/io/debezium/transforms/outbox/EventRouter.java [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+
+import static io.debezium.transforms.outbox.EventRouterConfigDefinition.parseAdditionalFieldsConfig;
+import static org.apache.kafka.connect.transforms.util.Requirements.requireStruct;
+
+/**
+ * Debezium Outbox Transform Event Router
+ *
+ * @author Renato mefi (gh@mefi.in)
+ */
+public class EventRouter<R extends ConnectRecord<R>> implements Transformation<R> {
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/transforms/outbox/EventRouter.java]: @jpechane, can you add the new `@Incubating` annotation before mergin?

Comment (renatomefi) [reply debezium-core/src/main/java/io/debezium/transforms/outbox/EventRouter.java]: I've rebased with master and added the annotation! :)

**Comment by jpechane** [comment]
Superseded with https://github.com/debezium/debezium/pull/820

**Comment by renatomefi** [comment]
Merged via #820
